### ISSUE TITLE
Semaphore not being released in case of error in async composition using HystrixObservableCommand/Collapser

### ISSUE THREAD

**Issue by amitcse** [Netflix/Hystrix#901]
Hi,

I am facing an issue where isolation semaphores of `HystrixObservableCommand` is not being released if the observable composition fails somewhere in between. 

My observable composition using `HystrixObseravbleCommand` and `HystrixObservableCollapser` for async calls looks something like this:

```
Observable.from(id1, id2, id3, ....)
.flatMap(id -> asyncCall1(id))
.flatMap(asyncCall1Result -> {
	asyncCall2Result -> asyncCall2(asyncCall1Result);
	asyncCall3Result -> asyncCall3(asyncCall1Result);
	asyncCall4Result -> asyncCall4(asyncCall1Result);

	return Observable.zip(asyncCall2Result, asyncCall3Result, asyncCall4Result, 
		(asyncCall2Result, asyncCall3Result, asyncCall4Result) 
					-> makeFinalResult(asyncCall2Result, asyncCall3Result, asyncCall4Result));
})
.map(asyncFinalResult -> doSomeTransformations(asyncFinalResult))
.toList().toBlocking().single();
```

The issue is if suppose *asyncCall2* fails after emitting some items, and there are in flight *asyncCall3* network calls, these semaphores used for *asyncCall3* (which uses `HystrixObservableCommand`) are never released.  If the same thing keeps happening, and all un-released semaphores equals `hystrix.command.default.fallback.isolation.semaphore.maxConcurrentRequests`, no request are able to be executed.

**Comment by mattrjacobs** [comment]
@amitcse In the code above, can you be specific on which `asyncCallX`s are using Hystrix, and if they are HystrixObservableCommands/HystrixCommands/HystrixObservableCollapsers/HystrixCollapsers?

**Comment by amitranjan-fk** [comment]
@mattrjacobs All the `asyncCallX` uses `HystrixObservableCollapser` under the hood. I think maybe I rectify this issue by doing a proper error handling, but not sure how to go about that.

**Comment by amitcse** [comment]
@mattrjacobs Here, all the `asyncCallX` are using `HystrixObservableCollapser` under the hood. I think maybe this issue can be rectified by doing a proper error handling, but not sure how to go about that.

**Comment by mattrjacobs** [comment]
Thanks for the info.  I'll see what kind of unit test coverage we already have for this scenario, and see if I can add to it.

**Comment by amitcse** [comment]
@mattrjacobs Turns out, the issue I am facing is because of the way how RxJava operator `zip` works. `Zip' is unsubscribing from the observable before onCompleted is executed, and hence semaphore is not released. Find the details [here] (https://github.com/ReactiveX/RxJava/issues/3124) on RxJava issues page. 

The actual command where I was facing the issue was not called via a collapser, and calling it via `HystrixObservableCollapser` did solve the issue for me. I was not using `HystrixObserableCollapser` because the dependent system doesn't support batch request, and I have to use a batch size of 1 as of now.

I think calling `observe` instead of `toObservable` on `HystrixObservableCommand` will also solve this issue, and I did test that semaphore were releasing when I called `observe` on `HystrixObservableCommand` , but I am not 100% sure of any other side effect that is may cause.
